src/wallet/wallet2.cpp

```diff
@@ -9035,18 +9059,21 @@ void wallet2::get_outs(std::vector<std::vector<tools::wallet2::get_outs_entry>>
         }
       }
 
-      // then pick others in random order till we reach the required number
-      // since we use an equiprobable pick here, we don't upset the triangular distribution
-      std::vector<size_t> order;
-      order.resize(requested_outputs_count);
-      for (size_t n = 0; n < order.size(); ++n)
-        order[n] = n;
-      std::shuffle(order.begin(), order.end(), crypto::random_device{});
-
+      // While we are still lacking outputs in this result ring, in our secret pick order...
       LOG_PRINT_L2("Looking for " << (fake_outputs_count+1) << " outputs of size " << print_money(td.is_rct() ? 0 : td.amount()));
-      for (size_t o = 0; o < requested_outputs_count && outs.back().size() < fake_outputs_count + 1; ++o)
+      for (size_t ring_pick_idx = base; ring_pick_idx < base + requested_outputs_count && outs.back().size() < fake_outputs_count + 1; ++ring_pick_idx)
       {
-        size_t i = base + order[o];
+        const get_outputs_out attempted_output = secret_picking_order[ring_pick_idx];
```

Comment:
I don't see how this is an improvement - if a dummy index is selected twice, why would selecting the earlier index matter (when they should be identical)?

This change should have basically no observable effects on the behavior, but does add an unnecessary `std::vector` to the processing loop.

Reply:
What do you mean by "dummy index"?

Reply:
Each output is referenced by index within an amount value. A dummy index is an output (index) that is not the "real" spend.

Reply:
Ah okay, I personally use the term decoy, but that's unimportant. 

> if a dummy index is selected twice, why would selecting the earlier index matter (when they should be identical)?

That's the thing, any decoy index is never selected twice because we specifically choose not to do that; see the variable `seen_indices`. We only ever choose *unique* decoy indexes within any given ring, and more than the current ring size at that, assuming that some will be invalid. Since we are ever only choosing unique decoy index values, each successive decoy index value depends upon all previous decoy index values. See this analogy:

> ... imagine that you have 1023 marbles. You label 512 marbles with the letter A, label 256 with the letter B, so on and so forth, finally labelling one marble with the letter J. You put them all into a bag, shake it well, and pick 8 marbles from the bag, but everytime you pick a marble of a certain letter, you remove all the other marbles from that bag with the same letter. That very first pick, the odds of picking a certain marble are exactly how you would expect: you are twice as likely to pick A as you are B, twice as likely to pick B as you are C, etc. However, on the second pick, the odds of getting the first pick are 0%, and the chances for everything else is higher. As you go down the line, your picked marbles will have letters that are increasingly more unlikely to pick if you hadn't removed the other marbles. In other words, the distribution of the later marbles will be more "skewed" in comparison to your original distribution of marbles.

If you perform this experiment multiple times, you will notice that you get "too many" J, H, & I marbles compared to if you picked with replacement. This effect gets worse towards the end since you are exhausting choices for each successive pick. For the first pick you have 10 choices, then 9, then 8, then 7, etc, until the 10th choice when you have picked exactly one marble for each letter A-J. This sampling is extremely different from the sampling of marbles initially in the bag.

In our case, the initial sampling of marbles in the bag represents our ideal target distribution. But whenever we use `seen_indices` to enforce unique values for our picks, we start warping the distribution towards a uniform distribution the longer we pick. As of writing this comment, there are 82037944 RingCT outputs on-chain. If we were to pick 82037944 *unique* RingCT outputs, we would end up with a distribution that looks nothing like a gamma distribution. If we picked 40 million unique RingCT outputs, it also would look nothing like a gamma distribution. However, the picks at the front of the list would follow the distribution more closely since they have fewer value constraints.

> This change should have basically no observable effects on the behavior

It does have a small, but observable effect

Reply:
B/c discarding duplicates measurably changes the distribution the more and more you pick. Would it be beneficial to empirically show this observation by releasing my source code for the aggregate decoy selection distribution?

Reply:
No, I forgot about the "overselection" portion. If there was no overselection, then this wouldn't be doing anything.

Reply:
There's still measurable dependence even if we don't over-select, but there's nothing we can do about that unless we want to hard-fork to allow duplicate ring members.